verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,402 @@
+
+    The set of records that each chain can use as a valid local connection
+    end. For each chain, this set contains one record, since we are
+    modeling a single connection in this specification.
+ 
+ ***************************************************************************)
+ChainAConnectionEnds == 
+    [
+        connectionID : { "connAtoB" },
+        clientID : { "clientOnAToB" }
+    ]
+ChainBConnectionEnds ==
+    [
+        connectionID : { "connBtoA" },
+        clientID : { "clientOnBToA" }
+    ]
+
+AllConnectionEnds ==
+    ChainAConnectionEnds \union ChainBConnectionEnds
+
+AllClientIDs == 
+    { x.clientID : x \in AllConnectionEnds }
+
+AllConnectionIDs ==
+    { x.connectionID : x \in AllConnectionEnds }
+
+
+(* Bundle with variables that chain A has access to. *)
+chainAVars == <<inBufChainA,    (* Input message buffer. *)
+                outBufChainA,   (* Output message buffer. *)
+                storeChainA>>   (* The local chain store. *)
+
+(* Bundle with variables that chain B has access to. *)
+chainBVars == <<inBufChainB,    (* Input message buffer. *)
+                outBufChainB,   (* Output message buffer. *)
+                storeChainB>>   (* Local chain store. *)
+
+(* All variables specific to chains. *)
+chainStoreVars == <<storeChainA, storeChainB>>
+
+allVars == <<chainStoreVars, 
+             inBufChainA, inBufChainB, 
+             outBufChainA, outBufChainB,
+             maliciousEnv>>
+
+
+INSTANCE ICS3Types
+
+chmA == INSTANCE ConnectionHandshakeModule
+        WITH MaxChainHeight <- MaxHeight,
+             inBuf          <- inBufChainA,
+             outBuf         <- outBufChainA,
+             store          <- storeChainA,
+             ConnectionIDs  <- { x.connectionID : x \in ChainAConnectionEnds },
+             ClientIDs      <- { x.clientID : x \in ChainAConnectionEnds }
+
+
+chmB == INSTANCE ConnectionHandshakeModule
+        WITH MaxChainHeight <- MaxHeight,
+             inBuf          <- inBufChainB,
+             outBuf         <- outBufChainB,
+             store          <- storeChainB,
+             ConnectionIDs  <- { x.connectionID : x \in ChainBConnectionEnds },
+             ClientIDs      <- { x.clientID : x \in ChainBConnectionEnds }
+
+
+(***************************************************************************
+ Environment actions.
+ ***************************************************************************)
+
+
+(* Environment initialization.
+    
+    This action kick-stars the ICS3 protocol by assigning an ICS3MsgInit
+    msg to either of the two chains (or both).
+    
+    Initially, the environment is non-malicious. The environment starts
+    acting maliciously once the connection on both chains transitions out
+    of state "UNINIT" (the initials state). It is important to
+    initialize the protocol like this, otherwise the env. can provoke a
+    deadlock. This can happen with the following sequence of actions:
```

Comment:
I think that we don't want to constrain scenario that leads to deadlock (at least at the model level). We want to be able to point out to the problematic scenario. Good scenarios should then captured by (most probably) temporal properties.